File: server/internal/mcpserver/projections.go
package mcpserver

import (
	"tallyo/internal/graph/model"
	u "tallyo/internal/utils"
)

type leanAsset struct {
	ID         string                `json:"id"`
	AssetType  model.AssetType       `json:"assetType"`
	Identifier string                `json:"identifier"`
	Name       *string               `json:"name,omitempty"`
	Classifier model.AssetClassifier `json:"classifier"`
}

func mapAsset(asset *model.Asset) leanAsset {
	return leanAsset{ID: asset.ID.EncodedString(), AssetType: asset.AssetType, Identifier: asset.Identifier, Name: asset.Name, Classifier: asset.Classifier}
}

type leanAccount struct {
	ID        string  `json:"id"`
	Name      string  `json:"name"`
	Type      string  `json:"type"`
	Subtype   *string `json:"subtype,omitempty"`
	OwnerID   string  `json:"ownerId"`
	OwnerName string  `json:"ownerName"`
	Manual    bool    `json:"manual,omitempty"`
	Closed    bool    `json:"closed,omitempty"`
	Hidden    bool    `json:"hidden,omitempty"`
}

func mapAccount(a *model.Account) leanAccount {
	return leanAccount{
		ID:        a.ID.EncodedString(),
		Name:      a.Name,
		Type:      string(a.Type),
		Subtype:   a.Subtype,
		OwnerID:   a.Owner.ID.EncodedString(),
		OwnerName: a.Owner.Name,
		Manual:    a.Manual,
		Closed:    a.Closed,
		Hidden:    a.Hidden,
	}
}

func mapAccounts(accounts []*model.Account) []leanAccount {
	return u.Map(accounts, mapAccount)
}

type leanList[T any] struct {
	Items []T `json:"items"`
}

func mapList[S, T any](items []S, f func(S) T) *leanList[T] {
	return &leanList[T]{Items: u.Map(items, f)}
}

func mapAccountList(list *model.AccountList) *leanList[leanAccount] {
	return mapList(list.Items, mapAccount)
}

type leanAccountPayload struct {
	Account leanAccount `json:"account"`
}

func mapAccountPayload(a *model.Account) leanAccountPayload {
	return leanAccountPayload{Account: mapAccount(a)}
}

type leanCategoryRef struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

func mapCategoryRef(c *model.Category) leanCategoryRef {
	return leanCategoryRef{ID: c.ID.EncodedString(), Name: c.Name}
}

File: server-rs/src/mcpserver/projections.rs
use serde::Serialize;

use crate::{
    ids::{GlobalId, GlobalIdType},
    schema::{Account, AccountType, Asset, AssetClassifier, AssetType, Category, CategoryGroup, CategoryKind, Owner},
};

pub(super) fn encode(typ: GlobalIdType, id: i64) -> String {
    GlobalId::new(typ, id).encoded_string()
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct LeanAsset {
    pub(super) id: String,
    pub(super) asset_type: AssetType,
    pub(super) identifier: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub(super) name: Option<String>,
    pub(super) classifier: AssetClassifier,
}

pub(super) fn map_asset(asset: Asset) -> LeanAsset {
    LeanAsset {
        id: encode(GlobalIdType::Asset, asset.id),
        asset_type: asset.asset_type,
        identifier: asset.identifier,
        name: asset.name,
        classifier: asset.classifier,
    }
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct LeanAccount {
    pub(super) id: String,
    pub(super) name: String,
    pub(super) r#type: AccountType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub(super) subtype: Option<String>,
    pub(super) owner_id: String,
    pub(super) owner_name: String,
    #[serde(skip_serializing_if = "std::ops::Not::not")]
    pub(super) manual: bool,
    #[serde(skip_serializing_if = "std::ops::Not::not")]
    pub(super) closed: bool,
    #[serde(skip_serializing_if = "std::ops::Not::not")]
    pub(super) hidden: bool,
}

pub(super) fn map_account(account: Account) -> LeanAccount {
    LeanAccount {
        id: encode(GlobalIdType::Account, account.id),
        name: account.name,
        r#type: account.r#type,
        subtype: account.subtype,
        owner_id: encode(GlobalIdType::Owner, account.owner.id),
        owner_name: account.owner.name,
        manual: account.manual,
        closed: account.closed,
        hidden: account.hidden,
    }
}

pub(super) fn map_accounts(accounts: Vec<Account>) -> Vec<LeanAccount> {
    accounts.into_iter().map(map_account).collect()
}

#[derive(Debug, PartialEq, Serialize)]
pub(super) struct LeanList<T> {
    pub(super) items: Vec<T>,
}

pub(super) fn map_list<S, T>(items: Vec<S>, map: impl FnMut(S) -> T) -> LeanList<T> {
    LeanList {
        items: items.into_iter().map(map).collect(),
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub(super) struct LeanAccountPayload {
    pub(super) account: LeanAccount,
}

#[derive(Debug, PartialEq, Serialize)]
pub(super) struct LeanCategoryRef {
    pub(super) id: String,
    pub(super) name: String,
}

pub(super) fn map_category_ref(category: &Category) -> LeanCategoryRef {
    LeanCategoryRef {
        id: encode(GlobalIdType::Category, category.id),
        name: category.name.clone(),
    }
}

// Go returned these model types as-is; their generated Rust counterparts carry raw ids.
#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct LeanCategory {
    pub(super) id: String,
    pub(super) name: String,
    pub(super) emoji: String,
    pub(super) group_name: String,
    pub(super) group_emoji: String,
    pub(super) kind: CategoryKind,
    pub(super) sort_order: i32,
    #[serde(rename = "plaidPFC2Codes")]
    pub(super) plaid_pfc2_codes: Vec<String>,
}

pub(super) fn map_category(category: Category) -> LeanCategory {
    LeanCategory {
        id: encode(GlobalIdType::Category, category.id),
        name: category.name,
        emoji: category.emoji,
        group_name: category.group_name,
        group_emoji: category.group_emoji,
        kind: category.kind,
        sort_order: category.sort_order,
        plaid_pfc2_codes: category.plaid_pfc2_codes,
    }
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct LeanCategoryGroup {
    pub(super) id: String,
    pub(super) name: String,
    pub(super) emoji: String,
    pub(super) kind: CategoryKind,
    pub(super) categories: Vec<LeanCategory>,
}

pub(super) fn map_category_group(group: CategoryGroup) -> LeanCategoryGroup {
    LeanCategoryGroup {
        id: encode(GlobalIdType::CategoryGroup, group.id),
        name: group.name,
        emoji: group.emoji,
        kind: group.kind,
        categories: group.categories.into_iter().map(map_category).collect(),
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub(super) struct LeanOwner {
    pub(super) id: String,
    pub(super) name: String,
}

pub(super) fn map_owner(owner: Owner) -> LeanOwner {
    LeanOwner {
        id: encode(GlobalIdType::Owner, owner.id),
        name: owner.name,
    }
}
